### PR TITLE
[Terrafrom] Add rate limiting

### PR DESCRIPTION
**Pull request by phalbert**
# Description

**What** - This PR implements rate limiting logic in the Terraform Cloud integration to address frequent `429 Too Many Requests` warnings in the logs.

**Why** - We were experiencing excessive `429 Too Many Requests` warnings, indicating that our integration was exceeding Terraform Cloud's API rate limits. This was cluttering the logs and potentially impacting performance.

**How** - Implemented a more robust rate limiting mechanism in the `TerraformClient` class:
1. Added a `wait_for_rate_limit` method to enforce rate limits.
2. Implemented a [token bucket algorithm](https://medium.com/@sumit-s/token-bucket-algorithm-a-laymans-guide-to-rate-limiting-b1ced8d0ee8e#:~:text=The%20Token%20Bucket%20Algorithm%20is,requests%20a%20user%20can%20make.) to manage API request rates.
3. Added asyncio locks to ensure thread-safe rate limiting in asynchronous operations.
4. Updated the `send_api_request` method to use the new rate limiting logic.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

<h4> All tests should be run against the port production environment(using a testing org). </h4>

### Core testing checklist

- [x] Integration able to create all default resources from scratch
- [x] Resync finishes successfully
- [x] Resync able to create entities
- [x] Resync able to update entities
- [x] Resync able to detect and delete entities
- [x] Scheduled resync able to abort existing resync and start a new one
- [x] Tested with at least 2 integrations from scratch
- [x] Tested with Kafka and Polling event listeners
- [x] Tested deletion of entities that don't pass the selector

### Integration testing checklist

- [x] Integration able to create all default resources from scratch
- [x] Resync able to create entities
- [x] Resync able to update entities
- [x] Resync able to detect and delete entities
- [x] Resync finishes successfully
- [ ] If new resource kind is added or updated in the integration, add example raw data, mapping and expected result to the `examples` folder in the integration directory.
- [ ] If resource kind is updated, run the integration with the example data and check if the expected result is achieved
- [ ] If new resource kind is added or updated, validate that live-events for that resource are working as expected
- [ ] Docs PR link [here](#)

### Preflight checklist

- [x] Handled rate limiting
- [x] Handled pagination
- [x] Implemented the code in async
- [x] Support Multi account

## Screenshots

[Include screenshots of logs showing reduced 429 warnings, if available]

## API Documentation

- [Terraform Cloud API Rate Limiting Documentation](https://developer.hashicorp.com/terraform/cloud-docs/api-docs#rate-limiting)

### CODE REVIEW

File: integrations/terraform-cloud/main.py [hunk 1]
```diff
@@ -136,21 +131,20 @@ async def fetch_runs_for_workspace(
...
+        )
 
-            tasks = [fetch_runs_for_workspace(workspace) for workspace in workspaces]
-            for completed_task in asyncio.as_completed(tasks):
-                workspace_runs = await completed_task
+        for chunk in [
+            workspaces[i : i + CHUNK_SIZE]
+            for i in range(0, len(workspaces), CHUNK_SIZE)
+        ]:
+            chunk_results = await gather(*[fetch_runs_for_workspace(w) for w in chunk])
```
Comment (mk-armah) [inline integrations/terraform-cloud/main.py:143]: Is there a specific reason for replacing as_completed with gather ? waiting for all tasks within a chunk to complete before moving on to the next chunk appears to impede performance in this context. Please correct me.

File: integrations/terraform-cloud/client.py [hunk 2]
```diff
@@ -32,7 +40,30 @@ def __init__(self, terraform_base_url: str, auth_token: str) -> None:
...
+            if len(self.request_times) >= RATE_LIMIT_PER_SECOND:
+                wait_time = 1 - (current_time - self.request_times[0])
+                if wait_time > 0:
+                    logger.info(
+                        f"Rate limit reached, waiting for {wait_time:.2f} seconds"
+                    )
+                    await asyncio.sleep(wait_time)
+                self.request_times = self.request_times[1:]
+
+            self.request_times.append(current_time)
```
Comment (mk-armah) [inline integrations/terraform-cloud/client.py:66]: [aiolimiter](https://aiolimiter.readthedocs.io/en/latest/) ?

**Review by mk-armah — CHANGES_REQUESTED** [review]
Great work 👏🏽, left few comments. Also I was wondering if we could use aiolimiter to control requests easily since [terraform rate limit](https://developer.hashicorp.com/terraform/cloud-docs/api-docs#rate-limiting) is time bound

**Comment by phalbert** [comment]
close for fresh impl in https://github.com/port-labs/ocean/pull/1110 with [aiolimiter](https://pypi.org/project/aiolimiter/)

File: integrations/terraform-cloud/client.py [hunk 3]
```diff
@@ -23,6 +25,12 @@ class CacheKeys(StrEnum):
 PAGE_SIZE = 100
 
 
+# https://developer.hashicorp.com/terraform/cloud-docs/api-docs#rate-limiting
+RATE_LIMIT_PER_SECOND = 30
+RATE_LIMIT_BUFFER = 5  # Buffer to avoid hitting the exact limit
+MAX_CONCURRENT_REQUESTS = 10
```
[the remark this answers is not in the document]
Comment (phalbert) [reply integrations/terraform-cloud/client.py:31]: 10 is how many workspaces we will process concurrently when fetching runs (or enriching these workspaces).

**Pull request closed without merging**